### ISSUE TITLE
Classifier evaluation

### ISSUE THREAD

**Issue by lazywei** [github/linguist#2205]
Hi, I'd like to discuss the current training / testing evaluation for the classifier (naive bayesian) with you guys.

For now, the training is using the files under `samples/`, and the testing stage is using [public gist](https://api.github.com/gists/public). However, it seems the distributions of these two sources are quite different. To be more specific, I'm using this code as a proof-of-concept: https://gist.github.com/lazywei/5d91efc4855c8658bf4f (using [Scikit-Learn's Multinomial NB](http://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.MultinomialNB.html))

I tried two way:

1. Randomly split `samples/` to get testing data: https://gist.github.com/lazywei/5d91efc4855c8658bf4f#file-sk-py-L31-L33.
2. Download the public gist as testing data: https://gist.github.com/lazywei/5d91efc4855c8658bf4f#file-sk-py-L36-L51. This is exactly the same thing done by [`rake classifer:test`](https://github.com/github/linguist/blob/master/Rakefile#L127-L142)

For the first one, training accuracy and testing accuracy are, respectively:
```
0.669222343921 # training
0.35737704918  # testing
```
while in the second one, I got:
```
0.674876847291 # training
0.128491620112 # testing
```

I think this basically implies that the distribution of two sources (samples and public gist) are probably not similar to each other. (otherwise we should get similar testing accuracy in 1. and 2.)

----

In my opinion, in order to improve the classifier, we need to establish a more consistent evaluation. For example, we should decide which one (samples or public gist) is more close to the real-world data distribution. And we should then stick with that one for evaluation.

Do you have any idea on this issue?

Thanks.

**Comment by bkeepers** [comment]
It looks like those scripts don't do any tokenization of the samples. I wonder if that contributes to the difference in the gists vs. samples, so so much of a source file is data and not language constructs.

We definitely need to find a good source for training data. @arfon and I have talked about using samples with modelines from public projects on GitHub as one source of data. We've also talked about adding a feature to GitHub to allow users to correct mis-detections, which could feed back into the classifier.

**Comment by lazywei** [comment]
> It looks like those scripts don't do any tokenization of the samples. I wonder if that contributes to the difference in the gists vs. samples, so so much of a source file is data and not language constructs.

Oops, you're right .. I originally thought the `linguist/tokenizer.rb` only builds the bag-of-words model, which is why I'm doing [`CountVectorizer`](https://gist.github.com/lazywei/5d91efc4855c8658bf4f#file-sk-py-L29). I'll then update my script to using the tokenizer, and check if there is still big difference between public gist and samples.



> We definitely need to find a good source for training data. @arfon and I have talked about using samples with modelines from public projects on GitHub as one source of data.

If there is no any PR doing this (crawling samples with modelines), I think I can work on it.

> We've also talked about adding a feature to GitHub to allow users to correct mis-detections, which could feed back into the classifier.

Agree. This can help us build a more accurate training dataset.

**Comment by bkeepers** [comment]
> If there is no any PR doing this (crawling samples with modelines), I think I can work on it.

@arfon We can probably use some internal infrastructure to get this more efficiently. Thoughts?

**Comment by lazywei** [comment]
Oh, right... totally forgot you guys have internal access LOL.
-- Chih-Wei (Bert)
Sent using CloudMagic
On 週日, 3月 08, 2015 at 12:10 下午, Brandon Keepers <notifications@github.com> wrote:
If there is no any PR doing this (crawling samples with modelines), I think I can work on it.


@arfon We can probably use some internal infrastructure to get this more efficiently. Thoughts?

**Comment by lazywei** [comment]
Just curious, why not just using [public gist](https://api.github.com/gists/public) for training if public gist provide the exact language?

**Comment by bkeepers** [comment]
I'm not sure what the history is. We could definitely explore it.

One issue we'll need to figure out is if the training data has to have an open source license for us to use it. Most Gists are unlicensed. I can ask our legal team for advice.

**Comment by lazywei** [comment]
Just for your information: so after utilizing the tokenizer, I got a closer result for NB:

```
0.562859490551 # training acc.
0.346625766871 # testing acc.
```

while for linear svm:

```
0.961380443714 # training acc.
0.447852760736 # testing acc.
```

Two things observed here:

1. The gap between training and testing stage is smaller, but it's still there. Either grasping data from public gist or using rosetta code should be fine.
2. Linear SVM should perform better after fine-tuning. Moreover, by blending these two classifiers (or even more others), we should get a better result. So, exploring new algorithm should be worth, IMO.

**Comment by pawandeep2155** [comment]
Lazwei I read about Linear SVM and it is used to predict if one object belongs to any of the two classes.
So I think it could only be used when file extension suggest two languages. I think it won't be ideal strategy for replacing or combining with Bayesian Classifier.

**Comment by lazywei** [comment]
@pawandeep2155 you're right, SVM is originally designed to be a binary classifier. However, it's very common to use the simple trick (meta-algorithm) "one versus one" and "one versus all" to turn binary classifier to be able to classify multiple classes. So it won't be an issue for SVM to predict more than two classes. In fact, it is very common to use SVM for multi-class task.
On the other hand, as long as we have two diverse (or more than two) classifiers, we can always perform blending (or voting) to reduce the variance. This trick is also very useful to improve accuracy -- even uniform blending can give significant improvement.

**Comment by arfon** [comment]
Cleaning up old issues
